### ISSUE TITLE
Unable to inject KEYCLOAK_INSTANCE in UseGuards(AuthGuard)

### ISSUE THREAD

**Issue by railsstudent** [ferrerojosh/nest-keycloak-connect#104]
My use case is I cannot register AuthGuard globally. Therefore, I commented out  

```
providers: [
    // {
    //   provide: APP_GUARD,
    //   useClass: AuthGuard,
    // },
    // {
    //   provide: APP_GUARD,
    //   useClass: ResourceGuard,
    // },
    // {
    //   provide: APP_GUARD,
    //   useClass: RoleGuard,
    // },
  ],
```

in app module and modified ProductController to apply UseGuards decorator
`  @Get()
  @Public(false)
  @Scopes('View')
  @UseGuards(AuthGuard)
  async findAll() {
    return await this.service.findAll();
  }`

When I ran the example app, error were printed in console

Nest can't resolve dependencies of the AuthGuard (?, KEYCLOAK_CONNECT_OPTIONS, KEYCLOAK_LOGGER, Reflector). Please make sure that the argument KEYCLOAK_INSTANCE at index [0] is available in the ProductModule context.

Potential solutions:
- If KEYCLOAK_INSTANCE is a provider, is it part of the current ProductModule?
- If KEYCLOAK_INSTANCE is exported from a separate @Module, is that module imported within ProductModule?
  @Module({
    imports: [ /* the Module containing KEYCLOAK_INSTANCE */ ]
  })

Error: Nest can't resolve dependencies of the AuthGuard (?, KEYCLOAK_CONNECT_OPTIONS, KEYCLOAK_LOGGER, Reflector). Please make sure that the argument KEYCLOAK_INSTANCE at index [0] is available in the ProductModule context.

Potential solutions:
- If KEYCLOAK_INSTANCE is a provider, is it part of the current ProductModule?
- If KEYCLOAK_INSTANCE is exported from a separate @Module, is that module imported within ProductModule?
  @Module({
    imports: [ /* the Module containing KEYCLOAK_INSTANCE */ ]
  })

Please kindly advise.

**Comment by deeeed** [comment]
I get the same error trying to debug the package. Any branch where a fix is available?

**Comment by railsstudent** [comment]
@deeeed  I don't fix there is a fix because no one reported this error in github issues.

**Comment by ferrerojosh** [comment]
Duplicate of #104

**Comment by railsstudent** [comment]
@ferrerojosh  I can still reproduce the issue in product controller.  Should I report it to Nest team? Thanks.

**Comment by ferrerojosh** [comment]
> @ferrerojosh I can still reproduce the issue in product controller. Should I report it to Nest team? Thanks.

Can I see how you imported the module ?

**Comment by ferrerojosh** [comment]
The example app ran fine to me, have you tried clearing your `node_modules` ? It is already strange enough that only `KEYCLOAK_INSTANCE` is missing.

**Comment by ferrerojosh** [comment]
If what you are doing is not registering auth guard globally, you can do so by importing the `KeycloakConnectModule` in some common module of sorts. You can name it `AuthModule` or your `CoreModule` and import that to any modules that requires `AuthGuard`

**Comment by railsstudent** [comment]
Let me try it when I get back to work tomorrow. Thanks.

**Comment by railsstudent** [comment]
I created CoreModule,

```
import { Module } from '@nestjs/common';
import { KeycloakConnectModule } from 'nest-keycloak-connect';

@Module({
  imports: [KeycloakConnectModule]
})
export class CoreModule {}
```
Import CoreModule in AppModule and ProductModule

```
import { Module } from '@nestjs/common';
import { APP_GUARD } from '@nestjs/core';
import {
  AuthGuard, KeycloakConnectModule,
  ResourceGuard,
  RoleGuard
} from 'nest-keycloak-connect';
import { AppController } from './app.controller';
import { ConfigModule } from './config/config.module';
import { KeycloakConfigService } from './config/keycloak-config.service';
import { CoreModule } from './core/core.module';
import { ProductModule } from './product/product.module';

@Module({
  imports: [
    KeycloakConnectModule.registerAsync({
      useExisting: KeycloakConfigService,
      imports: [ConfigModule]
    }),
    CoreModule,
    ProductModule,
  ],
  providers: [ ],
  controllers: [AppController],
})
export class AppModule {}
```

```
import { Module } from '@nestjs/common';
import { ProductService } from './product/product.service';
import { ProductController } from './product/product.controller';
import { CoreModule } from '../core/core.module';

@Module({
  imports: [CoreModule],
  providers: [ProductService],
  controllers: [ProductController]
})
export class ProductModule {}
```

Add AuthGuard in ProductController

```
import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  Put,
  UseGuards,
} from '@nestjs/common';
import { AuthGuard } from 'nest-keycloak-connect';
import { Product } from './product';
import { ProductService } from './product.service';

@Controller('product')
export class ProductController {
  constructor(private service: ProductService) {}

  @Get(':code')
  @UseGuards(AuthGuard)
  async findByCode(@Param('code') code: string) {
    return await this.service.findByCode(code);
  }
}
```

The app still produces the same error.

**Comment by ferrerojosh** [comment]
You should register `KeycloakConnectModule` inside your `CoreModule`:

```typescript
import { Module } from '@nestjs/common';
import { ConfigModule } from '../config/config.module';
import { KeycloakConnectModule } from 'nest-keycloak-connect';
import { APP_GUARD } from '@nestjs/core';

@Module({
  imports: [
    KeycloakConnectModule.registerAsync({
      useExisting: KeycloakConfigService,
      imports: [ConfigModule]
    })],
  ],
})
export class CoreModule {}
```

Your `AppModule` won't need `KeycloakConnectModule` anymore as well (unless you're using guards in your `AppController`, in that case import `CoreModule` here as well):
```typescript
import { Module } from '@nestjs/common';
import { AppController } from './app.controller';
import { CoreModule } from './core/core.module';

@Module({
  imports: [
    ProductModule,
  ],
  controllers: [AppController],
})
export class AppModule {}
```


Since you aren't using Keycloak globally, there is no need to to use `APP_GUARD` and just use `@UseGuards(AuthGuard)` normally.

**Comment by railsstudent** [comment]
Let me try again and get back to you, thanks for the prompt response.

**Comment by railsstudent** [comment]
Works for me now. Thanks.
I need to export KeycloakConnectModule in CoreModule or the application cannot start.

```
import { Module } from '@nestjs/common';
import { KeycloakConnectModule } from 'nest-keycloak-connect';
import { ConfigModule } from '../config/config.module';
import { KeycloakConfigService } from '../config/keycloak-config.service';

@Module({
  imports: [    
    KeycloakConnectModule.registerAsync({
      useExisting: KeycloakConfigService,
      imports: [ConfigModule]
    })
  ],
  exports: [KeycloakConnectModule]
})
export class CoreModule {}
```

```
import { Module } from '@nestjs/common';
import { ProductService } from './product/product.service';
import { ProductController } from './product/product.controller';
import { CoreModule } from '../core/core.module';

@Module({
  imports: [CoreModule],
  providers: [ProductService],
  controllers: [ProductController]
})
export class ProductModule {}
```

```
import { Module } from '@nestjs/common';
import { AppController } from './app.controller';
import { ProductModule } from './product/product.module';

@Module({
  imports: [
    ProductModule,
  ],
  providers: [],
  controllers: [AppController],
})
export class AppModule {}
``

If I import CoreModule in other modules, will KeycloakConnectModule register more than once? Is KeycloakConnectModule singleton instance?

Do you need help to make a wiki for this use case?  It may be useful to other users with the same user case.

**Comment by ferrerojosh** [comment]
`KeycloakConnectModule` would be a single instance since it is only registered inside `CoreModule` and exported.

I will be closing this as you already find the issue resolved.

**Comment by sgentile** [comment]
I still get an error moving this to a separate module:

```
Error: Nest can't resolve dependencies of the KEYCLOAK_CONNECT_OPTIONS (?). Please make sure that the argument KeycloakConfigService at index [0] is available in the KeycloakConnectModule context.

Potential solutions:
- If KeycloakConfigService is a provider, is it part of the current KeycloakConnectModule?
- If KeycloakConfigService is exported from a separate @Module, is that module imported within KeycloakConnectModule?
  @Module({
    imports: [ /* the Module containing KeycloakConfigService */ ]
  })
  ```
